Refactor to use DateOnly and TimeOnly for date/time handling

axunonb opened the pull request:
Increased `CalDateTime` immutability

CalDateTime:
* `Date` is of type `DateOnly` (breaking vs. v4)
* Make `UtzTzId` a public string
* Remove `DateOnlyValue`
* Remove `HasDate` (always true) (breaking vs. v4)
* Remove `Millisecond` (breaking vs. v4)
* Remove `Ticks` (breaking vs. v4)
* Remove overload CalDateTime(int year, int month, int day, string? tzId) (breaking vs. v4)
* `HasTime` only has a getter (breaking vs. v4)
* `Time` is a getter of type `TimeOnly`. Values are always rounded to the nearest second (breaking vs. v4)
* `AddMilliseconds` takes `double` as argument, like `DateTime.AddMilliseconds` (breaking vs. v4)
* Remove `TimeOnlyValue`
* `Value` only has a getter (breaking vs. v4)
* `TzId` only has a getter (breaking vs. v4)
* Remove `AsSystemLocal` (breaking vs. v4)
* Remove `AsDateTimeOffset` (breaking vs. v4)
* Remove `AddMilliseconds` (breaking vs. v4)
* Remove `AddTicks` (breaking vs. v4)
* Remove unused `TimeSpan? operator -(CalDateTime? left, IDateTime? right)` (breaking vs. v4)
* Update Equality operators for comparing to floating date/time (breaking vs. v4)

IDateTime (breaking vs. v4):
* Remove `AsSystemLocal` (breaking vs. v4)`
* Remove `AsDateTimeOffset` (breaking vs. v4)`
* Remove `AddMilliseconds` (breaking vs. v4)
* Remove `AddTicks` (breaking vs. v4)
* Remove `Millisecond` (breaking vs. v4)
* Remove `Ticks` (breaking vs. v4)
* Add explicit bool IsFloating (implicit TzId == null)
* `Date` is a getter of type `DateOnly`
* `Time` is a getter of type `TimeOnly`
* `HasTime` only has a getter
* Remove `HasDate` (always true) (breaking vs. v4)
* `Value` only has a getter
* `TzId` only has a getter
* `ToTimeZone(string?)` argument is a NRT

Miscellaneous:
* CTORs with `DateTime` arguments persist. Actually they are also comfortable, and now consistently processed
* Keep `IDateTime` - maybe in another PR
* Transitioned from DateTime to DateOnly and TimeOnly across the codebase for improved date and time handling.
* Updated methods, properties, and tests to ensure compatibility with the new approach.
* Updated xmldoc

Resolves #656
Resolves #662

axunonb committed:
Refactor to use DateOnly and TimeOnly for date/time handling

Increased `CalDateTime` immutability

CalDateTime:
* `Date` is of type `DateOnly` (breaking vs. v4)
* Make `UtzTzId` a public string
* Remove `DateOnlyValue`
* Remove `HasDate` (always true) (breaking vs. v4)
* Remove overload CalDateTime(int year, int month, int day, string? tzId) (breaking vs. v4)
* `HasTime` only has a getter (breaking vs. v4)
* `Time` is a getter of type `TimeOnly`. Values are always rounded to the nearest second (breaking vs. v4)
* Remove `TimeOnlyValue`
* `Value` only has a getter (breaking vs. v4)
* `TzId` only has a getter (breaking vs. v4)
* Remove `AsSystemLocal` (breaking vs. v4)
* Remove unused `TimeSpan? operator -(CalDateTime? left, IDateTime? right)` (breaking vs. v4)
* Update Equality operators for comparing to floating date/time (breaking vs. v4)

IDateTime (breaking vs. v4):
* Remove `AsSystemLocal`
* Add explicit bool IsFloating (implicit TzId == null)
* `Date` is a getter of type `DateOnly`
* `Time` is a getter of type `TimeOnly`
* `HasTime` only has a getter
* Remove `HasDate`
* `Value` only has a getter
* `TzId` only has a getter
* `ToTimeZone(string?)` argument is a NRT

Miscellaneous:
* CTORs with `DateTime` arguments persist. Actually they are also comfortable, and now consistently processed
* Keep `IDateTime` - maybe in another PR
* Transitioned from DateTime to DateOnly and TimeOnly across the codebase for improved date and time handling.
* Updated methods, properties, and tests to ensure compatibility with the new approach.
* Updated xmldoc

axunonb committed:
Remove AsDateTimeOffset from IDateTime and CalDateTime

axunonb committed:
Floating date/time can convert to any timezone ID, keeping Value unchanged

Remove fallback to system's local timezone for floating date/time

axunonb commented on Ical.Net.Tests/CalDateTimeTests.cs:
```diff
@@ -189,58 +154,150 @@ public static IEnumerable ToStringTestCases()
 
-    [Test]
-    public void SetValue_AppliesSameRulesAsWith_CTOR()
+    [Test, TestCaseSource(nameof(DateTimeArithmeticTestCases))]
```
Full coverage for all arithmetic operations

axunonb commented on Ical.Net.Tests/CalDateTimeTests.cs:
```diff
@@ -189,58 +154,150 @@ public static IEnumerable ToStringTestCases()
-    [Test]
-    public void SetValue_LeavesExistingPropertiesUnchanged()
+    [Test, TestCaseSource(nameof(EqualityTestCases))]
+    public bool EqualityTests(Func<IDateTime, bool> operation)
```
Full coverage for all equality comparisons

axunonb commented on Ical.Net.Tests/ComponentTest.cs:
```diff
@@ -22,29 +22,4 @@ public void UniqueComponent1()
         Assert.That(evt.DtStamp, Is.Not.Null);
     }
-
-    [Test, Category("Components")]
```
Value has no setter now

axunonb commented on Ical.Net.Tests/ProgramTest.cs:
```diff
@@ -52,8 +52,8 @@ public void Merge1()
 
         // Get occurrences for the first event
         var occurrences = evt1.GetOccurrences(
-            new CalDateTime(1996, 1, 1, _tzid),
```
Resharper auto-refactoring for changed signature

axunonb commented on Ical.Net/DataTypes/IDateTime.cs:
```diff
@@ -10,129 +10,135 @@ namespace Ical.Net.DataTypes;
     IDateTime AddMinutes(int minutes);
     IDateTime AddSeconds(int seconds);
-    IDateTime AddMilliseconds(int milliseconds);
+    IDateTime AddMilliseconds(double milliseconds);
```
same as `DateTilme.AddMilliseconds` - `int` is not useful

axunonb commented on Ical.Net/Evaluation/RecurrencePatternEvaluator.cs:
```diff
@@ -988,6 +990,6 @@ private static IDateTime MatchTimeZone(IDateTime dt1, IDateTime dt2)
 
         return dt1.IsUtc
             ? new CalDateTime(copy.AsUtc)
-            : new CalDateTime(copy.AsSystemLocal);
```
Isn't the code in `MatchTimeZone(IDateTime dt1, IDateTime dt2)` completely useless?

minichma replied:
Its not the most elegant code, but it aims to adjust UNTIL, which usually is specified in UTC, to DTSTART's time zone, which usually isn't in UTC. This allows checking comparing the adjusted UNTIL with the generated recurrences without considering the time zone again. So the functionality is needed one way or the other.

axunonb replied:
That's what I thought, but the way it is implemented looks like a noop. Then I removed the method invocation temporarily with no effect on unit tests. So this is just a hint re recurrences.

minichma replied:
I could easily be wrong, or relevant tests are missing. We'd need a test with UNTIL exactly at the last recurrence and then run the test with a TZ with positive offset and with one with a negative offset.

axunonb replied:
The 2 arguments to `MatchTimeZone(IDateTime dt1, IDateTime dt2)` always have the same timezone, so there is nothing to adjust?
```C#
if (r.Until != DateTime.MinValue)
{
    r.Until = MatchTimeZone(referenceDate, new CalDateTime(r.Until, referenceDate.TzId)).Value;
}
```

minichma replied:
You are fully right, its a noop. But I strongly believe that something's wrong there, because the conversion needs to be done somewhere. Should be investigated separately. Maybe the translation between UTC and tzid happened in the constructor of CalDateTime and was lost in one of the previous PRs?

axunonb replied:
Created #663 to track

axunonb committed:
Added unit test GetOccurrenceShouldExcludeDtEndZoned

Resolves #660

axunonb committed:
Fix: Initialize CalDateTime with timezone UTC only, if TimeOnly is not null

Reasoning: DATE cannot have a timezone

axunonb commented on Ical.Net/Calendar.cs:
```diff
@@ -238,11 +244,17 @@ public virtual HashSet<Occurrence> GetOccurrences(DateTime startTime, DateTime e
 
     /// <inheritdoc cref="GetOccurrences(IDateTime)"/>
     public virtual HashSet<Occurrence> GetOccurrences<T>(DateTime dt) where T : IRecurringComponent
-        => GetOccurrences<T>(new CalDateTime(dt.Date), new CalDateTime(dt.Date.AddDays(1).AddTicks(-1)));
```
This is probably meant to be `TimeSpan.FromSeconds(-1)` because 1 tick just gets truncated?

minichma commented on Ical.Net/Calendar.cs:
```diff
@@ -238,11 +244,17 @@ public virtual HashSet<Occurrence> GetOccurrences(DateTime startTime, DateTime e
     /// <param name="dt">The date for which to return occurrences. Time is ignored on this parameter.</param>
     /// <returns>A list of Periods representing the occurrences of this object.</returns>
     public virtual HashSet<Occurrence> GetOccurrences<T>(IDateTime dt) where T : IRecurringComponent
-        => GetOccurrences<T>(new CalDateTime(dt.Date), new CalDateTime(dt.Date.AddDays(1).AddTicks(-1)));
```
Now that CalDateTime is rounded to seconds, ticks aren't helpful anymore. Question would be generally, whether subtracting an 'Epsilon' is the best idea. It would certainly be cleaner to get the comparisons later on right, i.e. to compare for endTime to be exclusive, rather than inclusive.

minichma commented:
Just noted, that I didn't submit my review comments from 2d ago, so please read my comments in the context of last Sat.

minichma commented:
This is a great improvement regarding readability and maintainability!

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -228,35 +253,54 @@ public override int GetHashCode()
+            return false;
+        }
+
+        if (left.IsFloating || right.IsFloating)
```
On the first glance I thought the proposal worked, but it doesn't.

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -154,29 +174,36 @@ public CalDateTime(DateOnly date, TimeOnly? time, string? tzId = null)
-        HasDate = true;
         _dateOnly = dateOnly;
-        _timeOnly = timeOnly ?? new TimeOnly();
+        _timeOnly = RoundTimeToNearestSecond(timeOnly);
```
> keep the original value

Before this centralized rounding there was a mixture, see e.g. `DateUtil.Truncate` and other adjustments that did not use this method. I think the general limitation to seconds brings advantages in maintenance and reproducibility. I agree truncating is more consequent. Will modify.

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -486,126 +509,139 @@ TzId is null
+    /// This will also add a <see cref="IDateTime.Time"/> part that did not exist before the operation,
     /// if the seconds are not a multiple of 86400.
     /// </remarks>
     public IDateTime AddSeconds(int seconds)
```
Happy to hear your suggestion, as this was my first conclusion, too. It's one of "guess the users intention".
Let's throw an `InvalidOperationException`. Will modify.

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -486,126 +509,139 @@ TzId is null
     /// Milliseconds less than full seconds get truncated.
     /// </remarks>
-    public IDateTime AddMilliseconds(int milliseconds)
+    public IDateTime AddMilliseconds(double milliseconds)
```
Yep, will remove milliseconds and ticks

axunonb commented on Ical.Net/Evaluation/RecurrenceUtil.cs:
```diff
@@ -13,8 +14,12 @@ namespace Ical.Net.Evaluation;
-        new CalDateTime(dt.Date), new CalDateTime(dt.Date.AddDays(1).AddSeconds(-1)), includeReferenceDateInResults);
+    public static HashSet<Occurrence> GetOccurrences(IRecurrable recurrable, IDateTime dt, bool includeReferenceDateInResults)
+    {
+        var endTimeOnly = dt.Time?.Add(TimeSpan.FromSeconds(-1));
```
InvalidOperationException

minichma replied:
Not sure, I understand. I think the problem still exists. If someone specifies a dt e.g. 20241203T000000, which is legit, then this code would lead to 20241203 T -000001.

minichma replied:
I think, we would introduce a new bug here and should consider fixing it right away.

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -266,35 +310,45 @@ public override int GetHashCode()
     {
         var copy = left.Copy<CalDateTime>();
+        var newValue = copy.Value - right;
         if ((right.Ticks % TimeSpan.TicksPerDay) != 0)
```
Will correct

axunonb committed:
Refactor time handling and remove millisecond/ticks methods

* Updated CalDateTime to truncate time to seconds instead of rounding.
* Removed AddMilliseconds and AddTicks methods and related test cases.
* Fix: minus and plus operators for cases where time parts are equal
* Adjusted EndOfDay method to use AddSeconds(-1) instead of AddTicks(-1)
* Removed GetOccurrenceShouldExcludeDtEndZoned test.

minichma commented on Ical.Net/Utility/DateUtil.cs:
```diff
@@ -19,7 +19,7 @@ public static IDateTime StartOfDay(IDateTime dt)
         => dt.AddHours(-dt.Hour).AddMinutes(-dt.Minute).AddSeconds(-dt.Second);
 
     public static IDateTime EndOfDay(IDateTime dt)
```
We should probably get rid of that. Calculations should be done precisely, adding -1 whatever shouldn't be necessary.

axunonb replied:
Created issue #662 for a separate PR

minichma commented:
Sorry for the many comments. You're working on the big ones here, so many things to consider.

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -154,29 +174,37 @@ public CalDateTime(DateOnly date, TimeOnly? time, string? tzId = null)
-        _tzId = string.Equals(UtcTzId, tzId, StringComparison.OrdinalIgnoreCase) ? UtcTzId : tzId;
+        _tzId = tzId switch
+        {
+            _ when !timeOnly.HasValue && string.Equals(UtcTzId, tzId, StringComparison.OrdinalIgnoreCase) => UtcTzId,
```
thanks, that's redundant

axunonb commented on Ical.Net/CalendarComponents/CalendarEvent.cs:
```diff
@@ -152,15 +152,19 @@ public virtual bool IsAllDay
         set
         {
-            // Set whether or not the start date/time
+            // Set whether the start date/time
```
I just changed the wording of the comment :)

But yes, this one was easy. I removed the setter that was used only in 1 unit test. `AllDay` can simply be set with a `CalDateTime` without time.

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -266,35 +311,45 @@ public override int GetHashCode()
+        }
+        else
+        {
+            copy._dateOnly = DateOnly.FromDateTime(newValue);
```
Created  #661 for tracking a separate PR

axunonb commented on Ical.Net/DataTypes/IDateTime.cs:
```diff
@@ -10,129 +10,133 @@ namespace Ical.Net.DataTypes;
-    /// Gets the ticks for this date/time value.
+    /// Gets the ticks that applies to the <see cref="Value"/>.
     /// </summary>
     long Ticks { get; }
```
The question is applicable to other properties, too: Day, Month, Year...
All can be retrieved from the `Value` property. We could clean-up, then this would bring any advantage. But not in this PR.

minichma replied:
Sure, my thought on Milliseconds and Ticks is just, that they are truncated starting with this PR, so providing getters for them could be confusing and considered inconsistent.

axunonb committed:
Remove setter from `CalendarEvent.AllDay`

* Made `IsAllDay` in `CalendarEvent.cs` read-only,
* Updated `Issue432_AllDay` unit test in `RecurrenceTests.cs` to remove time component from `Start` property and eliminate setting the `IsAllDay` property. (The only place where the setter was used.)
* Simplified `CalDateTime.cs` initialization by removing condition related to `UtcTzId` when `timeOnly` is not provided.

axunonb commented:
No need for excuses :) - comments are all very welcome. 
There is one more commit with 2 issues from your latest comments.
For other comments we have new issues to track them.
One important topic is the [discussion](https://github.com/ical-org/ical.net/discussions/606#discussioncomment-11451260) re comparing floating to zoned date/time, Let's resolve the topic there.

axunonb commented on Ical.Net/Calendar.cs:
```diff
@@ -205,11 +205,17 @@ public VTimeZone AddTimeZone(VTimeZone tz)
     public virtual HashSet<Occurrence> GetOccurrences(IDateTime dt)
-        => GetOccurrences<IRecurringComponent>(new CalDateTime(dt.Date), new CalDateTime(dt.Date.AddDays(1).AddSeconds(-1)));
+    {
+        var endTimeOnly = dt.Time?.Add(TimeSpan.FromSeconds(-1));
```
Track in #662

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -440,41 +435,53 @@ public string? TzId
+        {
+            return null;
+        }
+        var roundedSecond = time.Value.Second + (time.Value.Millisecond >= 500 ? 1 : 0);
```
Ah, this was not recovered after by Visual Studio crash - again: thanks

axunonb commented on Ical.Net/DataTypes/IDateTime.cs:
```diff
@@ -10,129 +10,133 @@ namespace Ical.Net.DataTypes;
-    /// Gets the millisecond for this date/time value.
+    /// Gets the millisecond that applies to the <see cref="Value"/>.
     /// </summary>
     int Millisecond { get; }
```
see below

axunonb committed:
Ensure transitive equality for CalDateTime

* `operator ==`: if (left.IsFloating != right.IsFloating) return false;
* Add unit test

axunonb committed:
Remove properties `Ticks` and `Millisecond`

axunonb commented:
@minichma PR comments should all all be resolved or addressed

minichma commented:
I'm on a business trip until Sat. Not sure, whether I'll find the time until then.

minichma commented on Ical.Net/Evaluation/RecurrencePatternEvaluator.cs:
```diff
@@ -954,8 +954,10 @@ public override HashSet<Period> Evaluate(IDateTime referenceDate, DateTime perio
 
-            referenceDate = referenceDate.Copy<IDateTime>();
-            referenceDate.HasTime = true;
+            if (referenceDate.HasTime)
```
This is a redundant check, because already checked in the previous line.

axunonb replied:
correct

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -428,53 +421,57 @@ public string? TzId
+    }
+
+    /// <inheritdoc/>
+    public IDateTime ToTimeZone(string otherTzId)
```
Yeah, I considered `null`, and finally decided for `otherTzId` `string` instead of `string?`, because `null` had the meaning "to floating". Well...

minichma replied:
Oh, didn't notice it's not nullable. I think it's good as is. Could add a method ToFloating at some point.

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -486,126 +483,93 @@ TzId is null
-        if (!dt.HasTime && hours % 24 > 0)
+        var copy = Copy<CalDateTime>();
+        var newValue = copy.Value.AddHours(hours);
+        if (!copy.HasTime && (hours % 24 == 0))
```
issues #661 and #662 already exists since last review

axunonb committed:
Remove redundant if clause

axunonb commented:
>  introduce a new bug here and should consider fixing it right away

Agree about this should be fixed (#662) with another PR.
It's existing code logic though:
https://github.com/ical-org/ical.net/blob/add58a9bc3dffb8014a59d0bc6a831dbfa5635f7/Ical.Net/Evaluation/RecurrenceUtil.cs#L16-L17

axunonb committed:
Simplify end period creation in GetOccurrences method

Removed the `endTimeOnly` variable and its usage in the `GetOccurrences` method. The code now directly creates the `CalDateTime` object for the end of the period without adjusting the time component. This change streamlines the code and avoids potential issues related to time adjustments.

axunonb commented:
Comments resolved or tracked by issues

axunonb committed:
Simplify GetOccurrences methods by removing time adjustments

The `GetOccurrences` methods in `Calendar.cs` have been simplified
by removing the code that adjusted the end time by subtracting one
second or one tick. The methods now directly use the start date
and the date one day after the start date without adjusting the
time component. This change simplifies the logic and removes
unnecessary time adjustments.

The adjustments are redudant, because
public static HashSet<Occurrence> RecurrenceUtil.GetOccurrences(IRecurrable recurrable, IDateTime periodStart, IDateTime periodEnd, bool includeReferenceDateInResults)
uses precise "LessThan" for comparing end date/times

Resolves  #662

axunonb commented:
The remaining time adjustments were only found in class `Calendar`.

The adjustments are redundant, because calls finally go to
`public static HashSet<Occurrence> RecurrenceUtil.GetOccurrences(IRecurrable recurrable, IDateTime periodStart, IDateTime periodEnd, bool includeReferenceDateInResults)`
which uses **precise** "LessThan" for comparing end date/times.

minichma commented:
> The fix was trivial, so it's done in https://github.com/ical-org/ical.net/commit/5956fb998d61514362f39407a3a5ac50bb13eb88

> It's existing code logic though:

Great, thank you! The difference to the previous code is, that the previous version calculated 23:59:59 the previous day, but the new code calculated -0:00:01 the same day. So now its certainly better.

minichma reviewed (approved):
Great improvement!

axunonb commented on Ical.Net/DataTypes/CalDateTime.cs:
```diff
@@ -488,13 +488,6 @@ public IDateTime ToTimeZone(string otherTzId)
-    /// If the <see cref="TzId"/> is <see langword="null"/> or does not represent an
-    /// IANA or other well-known timezone ID, the system's local timezone will be used.
-    /// </remarks>
-    public DateTimeOffset AsDateTimeOffset => DateUtil.ToZonedDateTimeLeniently(Value, TzId).ToDateTimeOffset();
```
@CAPCHIK The reasons for the redesign are given at the top of this PR. This was the decision at this time.
We're currently not yet happy about `CalDateTime` as it is now,. All timezone related methods should go into a separate class, maybe as extension methods.
Any changes to `CalDateTime` should happen after the final design is clear. We welcome every helping hand with ical.net.
The new ical.net v5 release reflects the current WIP for `CalDatetime`.

> I have to use the same logic to extract the timezone offset from NodaTime.

Yes, for v5.0.0

CAPCHIK replied:
Thanks for the explanation, I will try to find a way to help in the development.

Bluejanis commented:
Is there a migration documentation for this?
I updated the nuget package and now AsDateTimeOffset is gone for example.

axunonb commented:
> Is there a migration documentation for this?

Probably not as complete as you wished, but have a look at the [Migration Guide](https://github.com/ical-org/ical.net/wiki/Migrating-Guides) and the [API Changes](https://github.com/ical-org/ical.net/wiki/API-Changes-v4-to-v5)
